File: BackEnd/PHP/laravelshop/app/Modules/News/Services/NewsService.php
<?php

namespace App\Modules\News\Services;

use App\Common\Exceptions\BusinessException;
use App\Models\News;
use App\Models\NewsCategory;
use App\Models\NewsProduct;
use App\Models\Product;
use App\Models\User;
use Illuminate\Support\Str;
use Symfony\Component\HttpFoundation\Response;

class NewsService
{
    public function list(?int $categoryId, int $page, int $limit): array
    {
        $result = News::query()->when($categoryId !== null, fn ($q) => $q->where('category_id', $categoryId))->orderByDesc('id')->paginate($limit, ['*'], 'page', $page);

        return [
            'items' => $result->getCollection()->map(fn (News $news): array => $this->serializeNews($news))->all(),
            'pagination' => $this->pagination($result),
        ];
    }

    public function categories(): array
    {
        return NewsCategory::query()->orderBy('id')->get()->map(fn (NewsCategory $category): array => [
            'id' => $category->id,
            'name' => $category->name,
            'slug' => $category->slug,
        ])->all();
    }

    public function show(int $id): array
    {
        $news = News::query()->find($id);
        if (! $news) {
            throw new BusinessException('뉴스를 찾을 수 없습니다.', 'NEWS_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        $productIds = NewsProduct::query()->where('news_id', $news->id)->pluck('product_id');
        $products = Product::query()->whereIn('id', $productIds)->get()->map(fn (Product $product): array => [
            'id' => $product->id,
            'name' => $product->name,
            'slug' => $product->slug,
        ])->all();

        return array_merge($this->serializeNews($news), ['products' => $products]);
    }

    public function create(User $actor, array $payload): array
    {
        $this->assertAdmin($actor);
        $news = News::query()->create([
            'category_id' => $payload['categoryId'] ?? null,
            'author_id' => $actor->id,
            'title' => $payload['title'],
            'content' => $payload['content'],
            'thumbnail_url' => $payload['thumbnailUrl'] ?? null,
        ]);
        $this->syncProducts($news->id, $payload['productIds'] ?? []);

        return $this->show($news->id);
    }

    public function update(User $actor, int $id, array $payload): array
    {
        $this->assertAdmin($actor);
        $news = News::query()->find($id);
        if (! $news) {
            throw new BusinessException('뉴스를 찾을 수 없습니다.', 'NEWS_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        $news->forceFill([
            'category_id' => array_key_exists('categoryId', $payload) ? $payload['categoryId'] : $news->category_id,
            'title' => $payload['title'] ?? $news->title,
            'content' => $payload['content'] ?? $news->content,
            'thumbnail_url' => array_key_exists('thumbnailUrl', $payload) ? $payload['thumbnailUrl'] : $news->thumbnail_url,
        ])->save();

        if (array_key_exists('productIds', $payload)) {
            $this->syncProducts($news->id, $payload['productIds'] ?? []);
        }

        return $this->show($news->id);
    }

    public function delete(User $actor, int $id): array
    {
        $this->assertAdmin($actor);
        $news = News::query()->find($id);
        if (! $news) {
            throw new BusinessException('뉴스를 찾을 수 없습니다.', 'NEWS_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        $news->delete();

        return ['message' => '뉴스가 삭제되었습니다.'];
    }

    public function createCategory(User $actor, array $payload): array
    {
        $this->assertAdmin($actor);
        $category = NewsCategory::query()->create([
            'name' => $payload['name'],
            'slug' => Str::slug($payload['slug'] ?: $payload['name']),
        ]);

        return [
            'id' => $category->id,
            'name' => $category->name,
            'slug' => $category->slug,
        ];
    }

    public function deleteCategory(User $actor, int $id): array
    {
        $this->assertAdmin($actor);
        $category = NewsCategory::query()->find($id);
        if (! $category) {
            throw new BusinessException('뉴스 카테고리를 찾을 수 없습니다.', 'NEWS_CATEGORY_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        $category->delete();

        return ['message' => '뉴스 카테고리가 삭제되었습니다.'];
    }

    private function syncProducts(int $newsId, array $productIds): void
    {
        NewsProduct::query()->where('news_id', $newsId)->delete();
        foreach ($productIds as $productId) {
            if (Product::query()->where('id', $productId)->exists()) {
                NewsProduct::query()->create(['news_id' => $newsId, 'product_id' => $productId]);
            }
        }
    }

    private function serializeNews(News $news): array
    {
        return [
            'id' => $news->id,
            'categoryId' => $news->category_id,
            'authorId' => $news->author_id,
            'title' => $news->title,
            'content' => $news->content,
            'thumbnailUrl' => $news->thumbnail_url,
            'createdAt' => optional($news->created_at)?->toISOString(),
            'updatedAt' => optional($news->updated_at)?->toISOString(),
        ];
    }

    private function pagination($result): array
    {
        return [
            'page' => $result->currentPage(),
            'limit' => $result->perPage(),
            'total' => $result->total(),
            'totalPages' => $result->lastPage(),
        ];
    }

    private function assertAdmin(User $user): void
    {
        if ($user->role !== 'ADMIN') {
            throw new BusinessException('관리자 권한이 필요합니다.', 'FORBIDDEN', Response::HTTP_FORBIDDEN);
        }
    }
}

File: BackEnd/Java/java-spring-maven-jpa-postgresql/src/main/java/com/pbshop/java/spring/maven/jpa/postgresql/news/NewsService.java
package com.pbshop.java.spring.maven.jpa.postgresql.news;

import java.util.List;
import java.util.LinkedHashMap;
import java.util.Map;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.pbshop.java.spring.maven.jpa.postgresql.auth.security.AuthenticatedUserPrincipal;
import com.pbshop.java.spring.maven.jpa.postgresql.common.exception.BusinessException;
import com.pbshop.java.spring.maven.jpa.postgresql.common.exception.ErrorCode;
import com.pbshop.java.spring.maven.jpa.postgresql.news.dto.NewsDtos.CreateNewsCategoryRequest;
import com.pbshop.java.spring.maven.jpa.postgresql.news.dto.NewsDtos.CreateNewsRequest;
import com.pbshop.java.spring.maven.jpa.postgresql.news.dto.NewsDtos.UpdateNewsRequest;
import com.pbshop.java.spring.maven.jpa.postgresql.product.Product;
import com.pbshop.java.spring.maven.jpa.postgresql.product.ProductRepository;
import com.pbshop.java.spring.maven.jpa.postgresql.user.User;
import com.pbshop.java.spring.maven.jpa.postgresql.user.UserRepository;

@Service
@Transactional
public class NewsService {
    private final NewsRepository newsRepository;
    private final NewsCategoryRepository categoryRepository;
    private final NewsProductRepository newsProductRepository;
    private final ProductRepository productRepository;
    private final UserRepository userRepository;

    public NewsService(NewsRepository newsRepository, NewsCategoryRepository categoryRepository,
            NewsProductRepository newsProductRepository, ProductRepository productRepository, UserRepository userRepository) {
        this.newsRepository = newsRepository;
        this.categoryRepository = categoryRepository;
        this.newsProductRepository = newsProductRepository;
        this.productRepository = productRepository;
        this.userRepository = userRepository;
    }

    @Transactional(readOnly = true)
    public Map<String, Object> list(Long categoryId) {
        List<Map<String, Object>> items = (categoryId == null ? newsRepository.findAllByOrderByIdDesc() : newsRepository.findByCategoryIdOrderByIdDesc(categoryId))
                .stream().map(this::toSummary).toList();
        return Map.of("items", items, "pagination", Map.of("page", 1, "limit", items.isEmpty() ? 20 : items.size(), "total", items.size(), "totalPages", items.isEmpty() ? 0 : 1));
    }
    @Transactional(readOnly = true)
    public List<Map<String, Object>> categories() {
        return categoryRepository.findAllByOrderByIdAsc().stream().map(category -> {
            Map<String, Object> item = new LinkedHashMap<>();
            item.put("id", category.getId());
            item.put("name", category.getName());
            item.put("slug", category.getSlug());
            return item;
        }).toList();
    }
    @Transactional(readOnly = true)
    public Map<String, Object> show(Long id) {
        News news = requireNews(id);
        List<Map<String, Object>> products = newsProductRepository.findByNewsIdOrderByIdAsc(news.getId()).stream()
                .map(rel -> {
                    Map<String, Object> product = new LinkedHashMap<>();
                    product.put("id", rel.getProduct().getId());
                    product.put("name", rel.getProduct().getName());
                    product.put("slug", rel.getProduct().getSlug());
                    return product;
                }).toList();
        Map<String, Object> response = new LinkedHashMap<>();
        response.put("id", news.getId());
        response.put("categoryId", news.getCategory() == null ? null : news.getCategory().getId());
        response.put("authorId", news.getAuthor().getId());
        response.put("title", news.getTitle());
        response.put("content", news.getContent());
        response.put("thumbnailUrl", news.getThumbnailUrl() == null ? "" : news.getThumbnailUrl());
        response.put("products", products);
        response.put("createdAt", news.getCreatedAt() == null ? null : news.getCreatedAt().toString());
        response.put("updatedAt", news.getUpdatedAt() == null ? null : news.getUpdatedAt().toString());
        return response;
    }
    public Map<String, Object> create(AuthenticatedUserPrincipal principal, CreateNewsRequest request) {
        User actor = requireAdminUser(principal);
        News news = new News();
        news.setCategory(request.categoryId() == null ? null : categoryRepository.findById(request.categoryId()).orElse(null));
        news.setAuthor(actor);
        news.setTitle(request.title());
        news.setContent(request.content());
        news.setThumbnailUrl(request.thumbnailUrl());
        News saved = newsRepository.save(news);
        syncProducts(saved, request.productIds());
        return show(saved.getId());
    }
    public Map<String, Object> update(AuthenticatedUserPrincipal principal, Long id, UpdateNewsRequest request) {
        requireAdminUser(principal);
        News news = requireNews(id);
        if (request.categoryId() != null || request.categoryId() == null) { news.setCategory(request.categoryId() == null ? null : categoryRepository.findById(request.categoryId()).orElse(null)); }
        if (request.title() != null && !request.title().isBlank()) { news.setTitle(request.title()); }
        if (request.content() != null && !request.content().isBlank()) { news.setContent(request.content()); }
        if (request.thumbnailUrl() != null || request.thumbnailUrl() == null) { news.setThumbnailUrl(request.thumbnailUrl()); }
        if (request.productIds() != null) { syncProducts(news, request.productIds()); }
        return show(news.getId());
    }
    public Map<String, Object> delete(AuthenticatedUserPrincipal principal, Long id) { requireAdminUser(principal); newsRepository.delete(requireNews(id)); return Map.of("message", "뉴스가 삭제되었습니다."); }
    public Map<String, Object> createCategory(AuthenticatedUserPrincipal principal, CreateNewsCategoryRequest request) {
        requireAdminUser(principal);
        NewsCategory category = new NewsCategory();
        category.setName(request.name());
        category.setSlug((request.slug() == null || request.slug().isBlank() ? request.name() : request.slug()).toLowerCase().replace(" ", "-"));
        categoryRepository.save(category);
        return Map.of("id", category.getId(), "name", category.getName(), "slug", category.getSlug());
    }
    public Map<String, Object> deleteCategory(AuthenticatedUserPrincipal principal, Long id) { requireAdminUser(principal); categoryRepository.delete(categoryRepository.findById(id).orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "뉴스 카테고리를 찾을 수 없습니다."))); return Map.of("message", "뉴스 카테고리가 삭제되었습니다."); }
    private void syncProducts(News news, List<Long> productIds) {
        newsProductRepository.deleteByNewsId(news.getId());
        if (productIds == null) { return; }
        for (Long productId : productIds) {
            Product product = productRepository.findById(productId).orElse(null);
            if (product != null) {
                NewsProduct relation = new NewsProduct();
                relation.setNews(news);
                relation.setProduct(product);
                newsProductRepository.save(relation);
            }
        }
    }
    private News requireNews(Long id) { return newsRepository.findById(id).orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "뉴스를 찾을 수 없습니다.")); }
    private User requireAdminUser(AuthenticatedUserPrincipal principal) {
        if (principal == null) { throw new BusinessException(ErrorCode.UNAUTHORIZED, "인증이 필요합니다."); }
        if (!"ADMIN".equalsIgnoreCase(principal.role())) { throw new BusinessException(ErrorCode.FORBIDDEN, "관리자 권한이 필요합니다."); }
        return userRepository.findById(principal.userId()).orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "사용자를 찾을 수 없습니다."));
    }
    private Map<String, Object> toSummary(News news) {
        return Map.of("id", news.getId(), "categoryId", news.getCategory() == null ? null : news.getCategory().getId(), "authorId", news.getAuthor().getId(), "title", news.getTitle(), "content", news.getContent(), "thumbnailUrl", news.getThumbnailUrl() == null ? "" : news.getThumbnailUrl(), "createdAt", news.getCreatedAt() == null ? null : news.getCreatedAt().toString(), "updatedAt", news.getUpdatedAt() == null ? null : news.getUpdatedAt().toString());
    }
}
